Create login

StinaSoile opened the pull request:
I created
- login
- logout
- possibility to create new account
- saving progress for user.
Progress is not saved at all anymore if user is not signed in.

sign in:
![image](https://github.com/user-attachments/assets/94d95b6d-948f-4b43-8aa7-a66bdf2b915f)

create account:
![image](https://github.com/user-attachments/assets/592d1b1b-30f8-46d1-9804-cb83be1c3765)

StinaSoile committed:
feat: update author section text

StinaSoile committed:
feat: start making login form
- add login form component
- add username and password to state
- add userService with incomplete authenticateUser function

StinaSoile committed:
feat: make authenticateUser and handleLogin
- get token, username and progress from backend
- save them to localstorage

StinaSoile committed:
feat: make logout

StinaSoile committed:
feat: make create account

StinaSoile committed:
refactor: move states from CreateAccount to HomeScreen

StinaSoile committed:
feat: make error message if username is already in use
- make proper error message
- handle empty progress when saving users progress

StinaSoile committed:
chore: remove unneccessary comments and console.logs

sepi4 commented on src/components/HomeScreen/CreateAccount.tsx:
```diff
@@ -3,18 +3,97 @@ import {
 }: {
   handleClose: () => void;
   open: boolean;
+  handleCreateAccount: (e: React.FormEvent<HTMLFormElement>) => void;
```
voi ottaa ulos React objectista varmaan importissa?
`{Dispatch, SetStateAction }`

sepi4 commented on src/components/HomeScreen/CreateAccount.tsx:
```diff
@@ -3,18 +3,97 @@ import {
+  };
+
+  const validatePassword = () => {
+    if (newPassword.length < 8) {
```
sulut, yksi rivi. heti perässä ei sulkuja ja yksi rivi :)

sepi4 commented on src/components/HomeScreen/CreateAccount.tsx:
```diff
@@ -3,18 +3,97 @@ import {
+  };
+
+  const isErrors = () => {
+    if (
```
mitä if suluissa ( ) oleva juttu palauttaa? 
onko järkeä palauttaa true tai false?

sepi4 commented on src/components/HomeScreen/CreateAccount.tsx:
```diff
@@ -37,10 +116,90 @@ const CreateAccount = ({
+            value={newPassword}
+            onChange={(e) => {
+              setNewPassword(e.target.value);
+              if (passwordError) validatePassword();
```
hyviä noi onChange funktiot.

sepi4 commented on src/components/HomeScreen/CreateAccount.tsx:
```diff
@@ -37,10 +116,90 @@ const CreateAccount = ({
+          />
+          {isErrors() ? (
+            <Alert variant="outlined" severity="warning">
+              {duplicateNameError ? <>Nimi on jo käytössä. </> : <></>}
```
tossa voi käyttää varmaan && ternaryn sijasta
![image](https://github.com/user-attachments/assets/c97bb533-3f82-4c82-8094-300eefe0804b)

sepi4 commented on src/components/HomeScreen/HomeScreen.tsx:
```diff
@@ -4,23 +4,117 @@ import Login from "./Login";
+        await userService.createUser(newUsername, email, newPassword);
+        login(newUsername, newPassword);
+      } catch (error) {
+        if (error instanceof Error) {
```
miksi tämä tässä? onko tarpeellinen?

StinaSoile replied:
Muutin backendia, se palauttaa nyt 201 jos käyttäjä luodaan, ja 200 jos käyttäjänimi on käytössä. Eli ei enää palauta erroria. Poistin catchin, loin iffin. Onko ok?
![image](https://github.com/user-attachments/assets/325536a3-424e-474e-a990-281e23921dae)

sepi4 replied:
![image](https://github.com/user-attachments/assets/3fc8f77e-472b-40e0-b4d7-5e8903b58853)

sepi4 commented on src/components/HomeScreen/HomeScreen.tsx:
```diff
@@ -4,23 +4,117 @@ import Login from "./Login";
+
+  const handleCreateAccount = async (e: React.FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+    if (!emailError && !confirmError && !nameError && !passwordError) {
```
jos ei virheitä niin ei tehdä mitään tässä funktiossa. Mä tykkään laittaa että jos virhe niin `return`, näin poistuu yksi sisäkkäisyys taso.

sepi4 replied:
if (emailError || confirmError || nameError || passwordError) {
    return;
  }
  try {

sepi4 commented on src/services/userService.tsx:
```diff
@@ -0,0 +1,73 @@
+import { apiBaseUrl } from "../constants";
+import utils from "../utils";
+
+const authenticateUser = async (username: string, password: string) => {
```
tämä tiedosto ei varmaan tarvitse olla .tsx? voi varmaan olla .ts? tässä ei käytetä jsx:ää

sepi4 commented on src/utils.tsx:
```diff
@@ -0,0 +1,29 @@
+const parseAndValidateProgress = (
```
sama tässä ei näy jsx

sepi4 reviewed (changes requested):
Hyvä homma ✝
En ajanut omalla koneella mitään, katsoin vain koodia. Laitoin muutosehdotuksia joihinkin kohtiin jotka huomasin

StinaSoile committed:
refactor: make suggested changes
- change ternary to && in CreateAccount
- unneccessary tsx to ts
- remove unneccessary boolean return
- other suggested changes
